### DISCUSSION TITLE
Revert snapshot.json back to previous state before the latest migration:create

### DISCUSSION THREAD

**Discussion by khangphung-vl** [mikro-orm/mikro-orm#5878] [Q&A] [answered]
Hello, I've recently moved to Mikro ORM, Nest JS stack
and I face this problem over and over again in development process. I don't know whether to start a discussion here or on Umzug github page, I'm sorry if I'm wrong (I tried google search many times, but nothing found about this issue).

As I'm guessing, when the CLI command
```
npx mikro-orm migration:create
``` 
executed, it also updates the **.snapshot-abcd.json** with the current schema/entity relationship,... (you get my idea). 

**But what if the generated migration cannot run or never run (for example: wrong SQL syntax)?**
Now that we stuck with a wrong snapshot, even if I go back and correct the schema/entity, the **next** migration:create will just be all the **SQL alter, drop, add** call to a wrong schema (maybe never run because... it's wrong).

I knew a horrible solution was to drop the whole .snapshot-abcd.json and do the migration all over again. But that is not a proper way to solve it (definitely). So, how do you solve this problem?

Thank you.

**Comment by B4nan** [comment, marked as answer]
That's the whole point of the snapshot file, that it is updated based on created migrations, not executed ones. Why do you think the next migration should be wrong? It will be based on what you have in existing migrations. It sounds like you are removing the last migration, in such case you should also remove the snapshot file.

**Reply by khangphung-vl** [reply]
Yes, I did remove the last migration, because it was completely wrong SQL (I messed up with Mikro), and I don't want to change it manually. The execute of migrate:up also failed on that migration file.

**Reply by B4nan** [reply]
I see, maybe we should have an option to ignore the snapshot when creating a new migration for this case. But its effectively the same as if you remove the snapshot file, there is nothing wrong with that really.

**Reply by khangphung-vl** [reply]
Thank you so much, the rumors are true. You are indeed really nice 😄. I will close the discussion here.

**Reply by khangphung-vl** [reply]
In the case anyone wants to try the solution (delete everything and re-create the migration file), here's the command line for that, just to speed things up a bit, try to bind it into an npm/yarn command.
**Be careful, use as your own risk**
`find src/migrations -mindepth 1 -exec rm -rf {} + && npx mikro-orm migration:create`

Explain:
`-mindepth 1` means process all files **except** the starting-points (ie, the folder _migrations_ itself)

**Comment by vicpon** [comment]
I ran into this too. How do you remove a pending migration if it's added to the snapshot? The CLI should allow you to delete a migration and recreate the snapshot from the previous migration.

**Reply by B4nan** [reply]
Just remove the snapshot file too, it will be created again based on the current schema automatically. We can't "recreate snapshot based on previous migration", there is no previous state to revert to, migrations are just SQL queries (or even worse, some conditional statements, pretty much anything can be there), we'd have to analyze them one by one, that is a job for the database, not the ORM.

I can imagine a flag to ignore the snapshot, so you would remove the pending migration you don't like and run `migration create --ignore-snapshot`. Or we could have something like `migration create --ignore-pending` which would automatically remove all the pending migrations and the snapshot file.

**Reply by vicpon** [reply]
If I delete the snapshot and the unwanted migration file and undo the changes in the entities, how do I recreate the snapshot? If I try running migration:create again, it generates a migration that creates everything as if the tables had nothing in them.

The functionality I and OP mention exists in Entity Framework, which also keeps a database snapshot. The difference is that in EF migrations, you can remove a migration before it's applied to a database and it undoes those changes to the snapshot file as well.

**Reply by vicpon** [reply]
I just noticed the snapshot file is correct for the previous state. I guess the steps are to create a migration and delete it to restore the snapshot back to it's previous state. That works, though it's not intuitive

**Reply by B4nan** [reply]
> how do I recreate the snapshot

You don't, as I said above, it gets created automatically based on the current schema.
